src/middleware.ts

```diff
@@ -259,43 +302,62 @@ export const persist = <S extends State>(
   }
 
   // rehydrate initial state with existing stored state
-  ;(async () => {
-    const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
-
-    try {
-      const storageValue = await storage.getItem(name)
 
+  // a workaround to solve the issue of not storing rehydrated state in sync storage
+  // the set(state) value would be later overridden with initial state by create()
+  // to avoid this, we merge the state from localStorage into the initial state.
+  // TODO: find a better solution for this
+  let stateFromStorageInSync: S | undefined
+  const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
+  // bind is used to avoid `TypeError: Illegal invocation` error
+  toThenable(storage.getItem.bind(storage))(name)
+    .then((storageValue) => {
       if (storageValue) {
-        const deserializedStorageValue = await deserialize(storageValue)
-
-        // if versions mismatch, run migration
+        return deserialize(storageValue)
+      }
+    })
+    .then((deserializedStorageValue) => {
+      if (deserializedStorageValue) {
         if (deserializedStorageValue.version !== version) {
```

Comment:
```suggestion
        if ((deserializedStorageValue.version ?? false) && deserializedStorageValue.version !== version) {
```
see #409

Reply:
removing `?? false` just works?

Reply:
not if the version is `0`

Reply:
No, my bad. This doesn't even work if the version is `0`. So maybe something like: `typeof deserializedStorageValue.version === "number" && deserializedStorageValue.version !== version`?

Reply:
`StorageValue<S>` requires version to be `number` so we can even throw an error if typeof ... !== "number".

Reply:
Well, that wouldn't permit to store a primitive value as in #409
But maybe there is a reason not to allow that?

Reply:
well, if we want to allow non-version store, we should also change the `StorageValue<S>` type. I'm fine with either way.

Reply:
Yes, let's merge this first